From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivan Pekov <ivan@mrivanplays.com>
Date: Wed, 9 Sep 2020 16:49:11 +0300
Subject: [PATCH] Further improve tuinity's portal crash fix

so we never have equal directions if the direction block property isn't present on the given block data

diff --git a/src/main/java/net/minecraft/server/BlockPortalShape.java b/src/main/java/net/minecraft/server/BlockPortalShape.java
index a848447260f77b85e5f1bfaedeb16d7d068e00d8..993f58a2a0c168d705fc956e388c5a15c8bca31e 100644
--- a/src/main/java/net/minecraft/server/BlockPortalShape.java
+++ b/src/main/java/net/minecraft/server/BlockPortalShape.java
@@ -228,6 +228,16 @@ public class BlockPortalShape {
         BlockPosition blockposition = blockutil_rectangle.origin;
         IBlockData iblockdata = worldserver.getType(blockposition);
         EnumDirection.EnumAxis enumdirection_enumaxis1 = !iblockdata.contains(BlockProperties.E) ? enumdirection_enumaxis : (EnumDirection.EnumAxis) iblockdata.get(BlockProperties.E); // Tuinity - use provided direction if the block doesn't have one
+        // Yatopia start - improve tuinity's fix
+        // basically cycle thru the X and Z direction so we never have equal directions if we don't have that block property
+        if (!iblockdata.contains(BlockProperties.E)) {
+            if (enumdirection_enumaxis1 == EnumDirection.EnumAxis.X) {
+                enumdirection_enumaxis1 = EnumDirection.EnumAxis.Z;
+            } else if (enumdirection_enumaxis1 == EnumDirection.EnumAxis.Z || enumdirection_enumaxis == EnumDirection.EnumAxis.Y) {
+                enumdirection_enumaxis1 = EnumDirection.EnumAxis.X;
+            }
+        }
+        // Yatopia end
         double d0 = (double) blockutil_rectangle.side1;
         double d1 = (double) blockutil_rectangle.side2;
         int i = enumdirection_enumaxis == enumdirection_enumaxis1 ? 0 : 90;
